p2p/net/network.go

```diff
@@ -304,16 +307,21 @@ func (n *Net) accept(listen net.Listener) {
 		n.tcpSocketConfig(conn) // TODO maybe only set this after session handshake to prevent denial of service with big messages
 		c := newConnection(netConn, n, nil, nil, n.config.MsgSizeLimit, n.config.ResponseTimeout, n.logger)
 		go func(con Connection) {
-			defer func() { pending <- struct{}{} }()
-			err := c.setupIncoming(n.config.SessionTimeout)
-			if err != nil {
+			if err := c.setupIncoming(n.config.SessionTimeout); err != nil {
 				n.logger.Event().Warning("conn_incoming_failed", log.String("remote", c.remoteAddr.String()), log.Err(err))
 				return
 			}
-			go c.beginEventProcessing()
+
+			// Asynchronously handle the new connection
+			go func() {
+				// Return the new connection token once the connection is closed
+				defer func() { pending <- struct{}{} }()
```

Comment:
It has to live in the same goroutine as `beginEventProcessing()` - that's the source of the memory leak, that it would run before `beginEventProcessing()` finished. But, you're right, I will separately handle the case where `setupIncoming` failed, good catch.